pool-agent/cmd/agent.go

```diff
@@ -0,0 +1,319 @@
+	}
+	creatingInstances := make(map[string]instances)
+	for _, rt := range conf.ResourceTypesMap {
+		creatingInstances[rt.Name] = make(instances)
+	}
+
+	registry := prometheus.NewRegistry()
+	registry.Register(configuredInstancesCount)
+	registry.Register(lxdInstances)
+	for k, v := range conf.ResourceTypesCounts {
+		configuredInstancesCount.WithLabelValues(k).Set(float64(v))
+	}
+	agent := &Agent{
+		ImageAlias:     conf.ImageAlias,
+		InstanceSource: source,
+
+		ResourceTypesMap:    conf.ResourceTypesMap,
+		ResourceTypesCounts: conf.ResourceTypesCounts,
+		Client:              c,
+
+		CheckInterval:   checkInterval,
+		WaitIdleTime:    waitIdleTime,
+		ZombieAllowTime: zombieAllowTime,
+		currentImage: struct {
+			Hash      string
+			CreatedAt time.Time
+		}{Hash: "", CreatedAt: time.Time{}},
+
+		creatingInstances: creatingInstances,
+		deletingInstances: make(instances),
+		registry:          registry,
+	}
+	return agent, nil
+}
+
+func (a *Agent) reloadConfig() {
+	conf, err := LoadConfig()
+	if err != nil {
+		slog.Error("failed to reloadconfig", "err", err.Error())
+		return
+	}
+
+	for k, v := range conf.ResourceTypesCounts {
+		configuredInstancesCount.WithLabelValues(k).Set(float64(v))
+	}
+
+	if conf.ImageAlias != a.ImageAlias {
+		source, err := ParseImageAlias(conf.ImageAlias)
+		if err != nil {
+			slog.Error("parse image alias: %+v", "err", err.Error())
+			return
+		}
+		a.InstanceSource = source
+		a.ImageAlias = conf.ImageAlias
+	}
+	a.ResourceTypesMap = conf.ResourceTypesMap
+	a.ResourceTypesCounts = conf.ResourceTypesCounts
+}
+
+// Run runs the agent.
+func (a *Agent) Run(ctx context.Context, sigHupCh chan os.Signal) error {
+	ticker := time.NewTicker(a.CheckInterval)
+	defer ticker.Stop()
+
+	slog.Info("Started agent")
+
+	for {
+		select {
+		case <-sigHupCh:
+			slog.Info("Received SIGHUP. Reloading config...")
+			a.reloadConfig()
+		case <-ticker.C:
+			if err := a.checkInstances(); err != nil {
+				slog.Error("failed to check instances", "err", err.Error())
+			}
+			if err := prometheus.WriteToTextfile(metricsPath, a.registry); err != nil {
+				slog.Error("failed to write metrics: %+v", "err", err.Error())
+			}
+		case <-ctx.Done():
+			slog.Info("Stopping agent...")
+			return nil
+		}
+	}
+}
+
+func (a *Agent) countPooledInstances(instances []api.Instance, resourceTypeName string) int {
+	count := 0
+	for _, i := range instances {
+		if i.StatusCode != api.Frozen {
+			continue
+		}
+		if i.Config[configKeyImageAlias] != a.ImageAlias {
+			continue
+		}
+		if i.Config[configKeyResourceType] != resourceTypeName {
+			continue
+		}
+		if _, ok := i.Config[configKeyRunnerName]; ok {
+			continue
+		}
+		count++
+	}
+	return count
+}
+
+func generateInstanceName() (string, error) {
+	var b [4]byte
+	_, err := rand.Read(b[:])
+	if err != nil {
+		return "", fmt.Errorf("generate random id: %w", err)
+	}
+	return fmt.Sprintf("myshoes-runner-%x", b), nil
+}
+
+func (a *Agent) checkInstances() error {
```

Comment:
Please write comments that introduce `checkInstances` function.

- What is the check?
- Would it like to operate an instance?

Maybe you need to rename this function if no comments.